File: taxcalculator.py
import streamlit as st
import pandas as pd
import plotly.express as px
import re

# Set page configuration
st.set_page_config(page_title="AI Tax Assistant", page_icon="💰", layout="centered")

# Custom CSS for modern UI
st.markdown(
    """
    <style>
    body {
        background: linear-gradient(135deg, #F4F6F8, #E0E7FF);
        font-family: 'Inter', sans-serif;
        color: #333;
    }
    .main-title {
        font-size: 28px;
        font-weight: bold;
        text-align: center;
        margin-bottom: 10px;
        color: #1A73E8;
    }
    .container {
        background-color: white;
        padding: 20px;
        border-radius: 12px;
        box-shadow: 0px 4px 12px rgba(0, 0, 0, 0.1);
        margin: auto;
    }
    .stButton>button {
        background-color: #1A73E8;
        color: white;
        padding: 10px 20px;
        border-radius: 8px;
        font-size: 16px;
        width: 100%;
    }
    .stButton>button:hover {
        background-color: #1557B0;
    }
    </style>
    """,
    unsafe_allow_html=True
)

# Title
st.markdown(
    "<div class='main-title'>💸 Tax-O-Matic: Because Taxes Don't Calculate Themselves! 💸</div>",
    unsafe_allow_html=True
)

# Salary Input (Manual Entry)
st.subheader("Enter Your Annual Salary (₹)")
salary = st.number_input("Annual Salary (₹):", min_value=0, step=10000, format="%d")

# Dynamic Deductions (Based on Indian Tax Law)
st.subheader("Select Your Eligible Deductions")

deductions = {
    "Standard Deduction (₹50,000)": 50000,
    "EPF (Employee Provident Fund) - Max ₹1.5L": st.number_input("EPF Contribution (₹):", min_value=0, max_value=150000, step=5000),
    "PPF (Public Provident Fund) - Max ₹1.5L": st.number_input("PPF Contribution (₹):", min_value=0, max_value=150000, step=5000),
    "NPS (National Pension Scheme) - Max ₹50K": st.number_input("NPS Contribution (₹):", min_value=0, max_value=50000, step=5000),
    "ELSS (Equity Linked Savings) - Max ₹1.5L": st.number_input("ELSS Investment (₹):", min_value=0, max_value=150000, step=5000),
    "Home Loan Interest (Self-Occupied) - Max ₹2L": st.number_input("Home Loan Interest Paid (₹):", min_value=0, max_value=200000, step=5000),
    "Medical Insurance (80D) - Max ₹25K": st.number_input("Medical Insurance Premium (₹):", min_value=0, max_value=25000, step=5000),
    "Education Loan Interest (80E)": st.number_input("Education Loan Interest Paid (₹):", min_value=0, step=5000),
    "House Rent Allowance (HRA)": st.number_input("HRA Exemption (₹):", min_value=0, step=5000)
}

# Compute Total Deduction
total_deductions = sum(deductions.values())
st.write(f"💡 **Total Deductions Applied: ₹{total_deductions}**")

# Efficient Tax Calculation
def calculate_tax(income):
    """
    Calculate tax using Indian tax slabs.
    """
    tax_brackets = [
        (250000, 0.05),
        (500000, 0.10),
        (750000, 0.15),
        (1000000, 0.20),
        (1250000, 0.25),
        (1500000, 0.30)
    ]
    
    tax = 0
    previous_limit = 0

    for limit, rate in tax_brackets:
        if income > limit:
            tax += (limit - previous_limit) * rate
            previous_limit = limit
        else:
            tax += (income - previous_limit) * rate
            break
    else:
        tax += (income - previous_limit) * 0.30  # Tax for income above 1,500,000 at 30%

    return tax

# Calculate and display results
if st.button("🚀 Calculate Best Tax Option"):
    old_taxable_income = max(0, salary - total_deductions)
    new_taxable_income = salary  # New regime does not allow deductions

    old_tax = calculate_tax(old_taxable_income)
    new_tax = calculate_tax(new_taxable_income)
    
    best_option = "Old Regime" if old_tax < new_tax else "New Regime"

    st.subheader("Results")
    st.write(f"✅ **Old Tax Regime:** ₹{old_tax:.2f}")
    st.write(f"✅ **New Tax Regime:** ₹{new_tax:.2f}")
    st.success(f"🚀 Best Option: **{best_option}**")

    # Visualization
    st.subheader("Tax Comparison")
    df = pd.DataFrame({"Regime": ["Old Regime", "New Regime"], "Tax Amount": [old_tax, new_tax]})
    fig = px.bar(df, x="Regime", y="Tax Amount", text="Tax Amount", color="Regime",
                 title="Tax Comparison: Old vs New Regime",
                 labels={"Tax Amount": "Tax Amount (₹)", "Regime": "Tax Regime"})
    fig.update_traces(texttemplate='₹%{text:,.2f}', textposition='outside')
    fig.update_layout(showlegend=False)
    st.plotly_chart(fig, use_container_width=True)

    # AI-Based Tax Saving Suggestions
    st.subheader("💡 AI-Based Tax Saving Suggestions")
    if best_option == "Old Regime":
        suggestions = []
        if deductions["EPF (Employee Provident Fund) - Max ₹1.5L"] == 0:
            suggestions.append("✅ Consider investing in **EPF** to save up to ₹1.5L under Section 80C.")
        if deductions["PPF (Public Provident Fund) - Max ₹1.5L"] == 0:
            suggestions.append("✅ Investing in **PPF** can reduce your taxable income further.")
        if deductions["NPS (National Pension Scheme) - Max ₹50K"] == 0:
            suggestions.append("✅ **NPS investment** gives an extra ₹50,000 deduction under Section 80CCD(1B).")
        if deductions["Home Loan Interest (Self-Occupied) - Max ₹2L"] == 0:
            suggestions.append("✅ **Home loan interest** can provide ₹2L exemption under Section 24B.")
        if deductions["Medical Insurance (80D) - Max ₹25K"] == 0:
            suggestions.append("✅ Buying **medical insurance** saves tax under Section 80D.")
        for sug in suggestions:
            st.write(sug)
    else:
        st.write("🚀 The **New Regime** has lower tax rates but no exemptions. Invest wisely!")

File: TaxCalculator.js
import React, { useState } from "react";
import "./TaxCalculator.css";

const TaxCalculator = () => {
  const [salary, setSalary] = useState("");
  const [deductions, setDeductions] = useState({
    "Employee Provident Fund (EPF) - Max ₹1.5L": "",
    "National Pension System (NPS) - Max ₹50K": "",
    "Public Provident Fund (PPF)": "",
    "Life Insurance Premium (80C)": "",
    "Health Insurance (80D)": "",
    "Home Loan Interest (24B)": "",
    "Education Loan Interest (80E)": "",
    "Other Deductions": "",
    "Tax-saving Fixed Deposit": "",
    "Donations (80G)": "",
    "Interest on Savings Account (80TTA)": "",
    "Dividends Received": "",
    "Entertainment Deductions": "",
    "Parent Medical Expenses": "",
    "Travel Allowance": "",
    "Rent Paid": "",
  });
  const [result, setResult] = useState(null);
  const [analysis, setAnalysis] = useState("");

  const calculateOldTax = (income) => {
    let tax = 0;
    if (income <= 250000) return 0;
    else if (income <= 500000) tax = 0.05 * (income - 250000);
    else if (income <= 1000000) tax = 0.05 * 250000 + 0.2 * (income - 500000);
    else tax = 0.05 * 250000 + 0.2 * 500000 + 0.3 * (income - 1000000);
    return tax + tax * 0.04;
  };

  const calculateNewTax = (income) => {
    let tax = 0;
    if (income <= 250000) return 0;
    else if (income <= 500000) tax = 0.05 * (income - 250000);
    else if (income <= 750000) tax = 0.05 * 250000 + 0.1 * (income - 500000);
    else if (income <= 1000000)
      tax = 0.05 * 250000 + 0.1 * 250000 + 0.15 * (income - 750000);
    else if (income <= 1250000)
      tax =
        0.05 * 250000 + 0.1 * 250000 + 0.15 * 250000 + 0.2 * (income - 1000000);
    else
      tax =
        0.05 * 250000 +
        0.1 * 250000 +
        0.15 * 250000 +
        0.2 * 250000 +
        0.3 * (income - 1250000);
    return tax + tax * 0.04;
  };

  const handleCalculate = () => {
    const totalDeductions = Object.values(deductions).reduce(
      (acc, val) => acc + (Number(val) || 0),
      0
    );
    const taxableIncome = Math.max(0, salary - totalDeductions);

    const oldTax = calculateOldTax(taxableIncome);
    const newTax = calculateNewTax(taxableIncome);
    const bestOption = newTax < oldTax ? "New Regime" : "Old Regime";

    setResult({ oldTax, newTax, bestOption });

    setAnalysis(`
      <h4>Best Tax Option:</h4>
      <p>The best option for you is the <strong>${bestOption}</strong>.</p>
      <h4>Suggestions:</h4>
      ${
        bestOption === "Old Regime"
          ? `<p>Since the Old Regime benefits those with higher deductions, consider maximizing Section 80C investments like EPF, PPF, and NPS.</p>`
          : `<p>The New Regime is beneficial for those with fewer deductions. If you wish to save more, explore tax-efficient investment options outside 80C.</p>`
      }
    `);
  };

  return (
    <div className="tax-calculator">
      <h2>Tax Calculator</h2>
      <div className="input-section">
        <label>Annual Salary (₹)</label>
        <input
          type="number"
          placeholder="Enter salary"
          value={salary}
          onChange={(e) => setSalary(e.target.value)}
        />
      </div>

      <h3>Deductions</h3>
      <div className="deductions-container">
        {Object.keys(deductions).map((key) => (
          <div className="deduction-item" key={key}>
            <label>{key}</label>
            <input
              type="number"
              placeholder="₹0"
              value={deductions[key]}
              onChange={(e) =>
                setDeductions({ ...deductions, [key]: e.target.value })
              }
            />
          </div>
        ))}
      </div>

      <button className="calculate-btn" onClick={handleCalculate}>
        Calculate Tax
      </button>

      {result && (
        <div className="result-section">
          <h3>Results</h3>
          <p>
            <strong>Old Tax Regime:</strong> ₹{result.oldTax.toFixed(2)}
          </p>
          <p>
            <strong>New Tax Regime:</strong> ₹{result.newTax.toFixed(2)}
          </p>
          <p>
            <strong>Best Option:</strong> {result.bestOption}
          </p>
        </div>
      )}

      {analysis && (
        <div
          className="detailed-analysis"
          dangerouslySetInnerHTML={{ __html: analysis }}
        ></div>
      )}
    </div>
  );
};

export default TaxCalculator;
